sdk/src/main/scala/com/horizen/consensus/ConsensusDataProvider.scala

```diff
@@ -52,27 +59,52 @@ trait ConsensusDataProvider {
     FullConsensusEpochInfo(stakeEpochInfo, nonceEpochInfo)
   }
 
+  //Check possible stack buffer overflow situations
   def calculateNonceForEpoch(epochId: ConsensusEpochId): NonceConsensusEpochInfo = {
     val lastBlockIdInEpoch: ModifierId = lastBlockIdInEpochId(epochId)
     val lastBlockInfoInEpoch: SidechainBlockInfo = storage.blockInfoById(lastBlockIdInEpoch)
 
-    val nonceOpt: Option[BigInteger] = foldEpochRight[Option[BigInteger]](None, lastBlockIdInEpoch, lastBlockInfoInEpoch){
-      (blockId: ModifierId, blockInfo: SidechainBlockInfo, accumulator: Option[BigInteger]) =>
-        {
-          val minimalHashForCurrentBlockOpt: Option[BigInteger] = getMinimalHashOpt(blockInfo.mainchainBlockReferenceHashes.map(_.data))
-          (minimalHashForCurrentBlockOpt, accumulator) match {
-            case (None, _) => accumulator
-            case (minimalHashOpt, None) => minimalHashOpt
-            case (Some(a), Some(b)) => Option(a.min(b))
-        }
-      }
+    if (isGenesisBlock(lastBlockIdInEpoch)) {
+      ConsensusDataProvider.calculateNonceForGenesisBlockInfo(lastBlockInfoInEpoch)
     }
+    else {
+      calculateNonceForNonGenesisEpoch(lastBlockIdInEpoch, lastBlockInfoInEpoch)
+    }
+  }
 
-    require(nonceOpt.isDefined, "No mainchain reference had been found for whole consensus epoch") //crash whole world here?
-
-    NonceConsensusEpochInfo(bigIntToConsensusNonce(nonceOpt.get))
+  private def calculateNonceForNonGenesisEpoch(lastBlockIdInEpoch: ModifierId, lastBlockInfoInEpoch: SidechainBlockInfo): NonceConsensusEpochInfo = {
+    val allVrfOutputsWithSlots: List[(VrfProof, VrfProofHash, ConsensusSlotNumber)] =
```

Comment:
simplify all the nonce calculation also avoiding to have the full list of slots related infos (proofHash) that are involved in nonce calculation. Nonce calculation can be managed by updating the digest at each slot with the current proofHash. Moreover its not needed to loop in the first third of slots.